Spigot-API-Patches/0305-per-viewer-chat-rendering.patch

```diff
@@ -0,0 +1,356 @@
++            if(this.renderer != null) {
++                this.formatter = (displayName, message) -> this.renderer.render(this.player, displayName, message, this.legacyForwardingAudience());
++            } else {
++                requireNonNull(this.composer, "renderer, composer, and formatter");
++                this.formatter = (displayName, message) -> this.composer.composeChat(this.player, displayName, message);
++            }
+         }
+         return this.formatter;
+     }
+@@ -100,12 +187,13 @@ public abstract class AbstractChatEvent extends PlayerEvent implements Cancellab
+      *
+      * @param formatter the chat formatter
+      * @throws NullPointerException if {@code formatter} is {@code null}
+-     * @deprecated in favour of {@link #composer(ChatComposer)}
++     * @deprecated in favour of {@link #renderer(ChatRenderer)}
+      */
+     @Deprecated
+     public final void formatter(final @NotNull ChatFormatter formatter) {
+         this.formatter = requireNonNull(formatter, "formatter");
+-        this.composer = (source, displayName, message) -> formatter.chat(displayName, message);
++        this.composer = null;
++        this.renderer = null;
+     }
+ 
+     /**
+@@ -137,4 +225,13 @@ public abstract class AbstractChatEvent extends PlayerEvent implements Cancellab
+     public final void setCancelled(final boolean cancelled) {
+         this.cancelled = cancelled;
+     }
++
++    private @NotNull Audience legacyForwardingAudience() {
++        return new ForwardingAudience() {
++            @Override
++            public @NonNull Iterable<? extends Audience> audiences() {
++                return AbstractChatEvent.this.viewers;
++            }
++        };
++    }
+ }
+diff --git a/src/main/java/io/papermc/paper/event/player/AsyncChatEvent.java b/src/main/java/io/papermc/paper/event/player/AsyncChatEvent.java
+index a0f748957f4472103dd27fc95a711a42de7fae89..6fd76380dd855ba6639cf0d0130c6a12a98a7c0f 100644
+--- a/src/main/java/io/papermc/paper/event/player/AsyncChatEvent.java
++++ b/src/main/java/io/papermc/paper/event/player/AsyncChatEvent.java
+@@ -3,6 +3,8 @@ package io.papermc.paper.event.player;
+ import io.papermc.paper.chat.ChatComposer;
+ import io.papermc.paper.chat.ChatFormatter;
+ import java.util.Set;
++import io.papermc.paper.chat.ChatRenderer;
++import net.kyori.adventure.audience.Audience;
+ import net.kyori.adventure.text.Component;
+ import org.bukkit.entity.Player;
+ import org.bukkit.event.HandlerList;
+@@ -14,12 +16,20 @@ import org.jetbrains.annotations.NotNull;
+ public final class AsyncChatEvent extends AbstractChatEvent {
+     private static final HandlerList HANDLERS = new HandlerList();
+ 
++    public AsyncChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull Set<Audience> viewers, final @NotNull ChatRenderer renderer, final @NotNull Component message) {
```

Comment:
Do we want this constructor to take `Set<Player> recipients`, or just `Set<Audience> viewers`?